client/allocrunner/networking_cni.go

```diff
@@ -123,8 +158,201 @@ func (c *cniNetworkConfigurator) Setup(ctx context.Context, alloc *structs.Alloc
+			if tproxy.UID != "" {
+				proxyUID = tproxy.UID
+			}
+
+			// The value for the outbound Envoy port. The default value matches
+			// the default TransparentProxy service default for
+			// OutboundListenerPort. If the cluster admin sets this value to
+			// something non-default, they'll need to update the metadata on all
+			// the nodes to match. see also:
+			// https://developer.hashicorp.com/consul/docs/connect/config-entries/service-defaults#transparentproxy
+			if tproxy.OutboundPort != 0 {
+				proxyOutboundPort = int(tproxy.OutboundPort)
+			} else {
+				outboundPortAttr := c.nodeMeta["connect.transparent_proxy.default_outbound_port"]
+				parsedOutboundPort, err := strconv.ParseInt(outboundPortAttr, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf(
+						"could not parse default_outbound_port %q as port number: %w",
+						outboundPortAttr, err)
+				}
+				proxyOutboundPort = int(parsedOutboundPort)
+			}
+
+			// The inbound port is the service port exposed on the Envoy proxy
+			envoyPortLabel := "connect-proxy-" + svc.Name
+			if idx, ok := portLabels[envoyPortLabel]; ok {
+				proxyInboundPort = int(portMapping[idx].HostPort)
+			}
+
+			// Extra user-defined ports that get excluded from outbound redirect
+			if len(tproxy.ExcludeOutboundPorts) == 0 {
+				outboundPorts = nil
+			} else {
+				outboundPorts = helper.ConvertSlice(tproxy.ExcludeOutboundPorts,
+					func(p uint16) string { return fmt.Sprint(p) })
+			}
+
+			// The set of ports we'll exclude from inbound redirection
+			exposePortSet := set.From(exposePorts)
+
+			// We always expose reserved ports so that the allocation is
+			// reachable from the outside world.
+			for _, network := range tg.Networks {
+				for _, port := range network.ReservedPorts {
+					exposePortSet.Insert(fmt.Sprint(port.To))
+				}
+			}
+
+			// ExcludeInboundPorts can be either a numeric port number or a port
+			// label that we need to convert into a port number
+			for _, portLabel := range tproxy.ExcludeInboundPorts {
+				if _, err := strconv.ParseUint(portLabel, 10, 64); err == nil {
+					exposePortSet.Insert(portLabel)
+					continue
+				}
+				if idx, ok := portLabels[portLabel]; ok {
+					exposePortSet.Insert(
+						strconv.FormatInt(int64(portMapping[idx].ContainerPort), 10))
+				}
+			}
+
+			// We also exclude Expose.Paths. Any health checks with expose=true
+			// will have an Expose block added by the server, so this allows
+			// health checks to work as expected without passing thru Envoy
+			if svc.Connect.SidecarService.Proxy.Expose != nil {
+				for _, path := range svc.Connect.SidecarService.Proxy.Expose.Paths {
+					if idx, ok := portLabels[path.ListenerPort]; ok {
+						exposePortSet.Insert(
+							strconv.FormatInt(int64(portMapping[idx].ContainerPort), 10))
+					}
+				}
+			}
+
+			if exposePortSet.Size() == 0 {
+				exposePorts = nil
+			} else {
+				exposePorts = exposePortSet.Slice()
+				slices.Sort(exposePorts)
+			}
```

Comment:
If `exposePorts` start out as nil this can be 5 lines and a little easier to read.

And if you're brave enough to use `TreeSet` you don't need the sort at the end :slightly_smiling_face:

Reply:
Good catch on the `nil`. The overhead on the insert on the `TreeSet` doesn't seem worth it though, given that we're potentially inserting a handful of these?